### ISSUE TITLE
Implement optimal price controller, based on https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md

### ISSUE THREAD

**Issue by eskimor** [paritytech/polkadot#6643]

**Comment by rphmeier** [comment]
Reference Material for optimal controllers in a blockchain fee-setting:


https://research.web3.foundation/en/latest/polkadot/overview/2-token-economics.html#setting-transaction-fees
https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1559.md

**Comment by rphmeier** [comment]
> We will define an optimal bid queue size (depending on the number of parathread cores).

This could be configured by governance. If it is a storage item, the `Config` trait can expose a `ConfigureParathreadQueueOrigin` which can then be set by the runtime declaration.

**Comment by rphmeier** [comment]
I believe there needs to be a maximum bid queue size for ensuring that the runtime storage is not attackable. However, that may not be the case especially if the pallet is implemented with smart pagination techniques.

**Comment by eskimor** [comment]
> I believe there needs to be a maximum bid queue size for ensuring that the runtime storage is not attackable. However, that may not be the case especially if the pallet is implemented with smart pagination techniques.

Our approach so far would have been to make it economically bounded. Goal would be to make a price function that is relatively stable on low utilization but grows rather fast if the queue grows too big - in a way that makes resource exhaustion impossible, by requiring more DOTs than there are available to reach sizes that would become problematic.

This should work if practical limits are rather high, which I believe is the case. Alternatively we could also see whether we can re-use Gav's paging queue, which might be a good safety net if not too cumbersome, because we indeed also don't want the queue to grow too big for other reasons as well: We would like to maintain the property that an order that goes on chain, will be served soonish, because:

1. This makes for a good user interface.
2. We can then treat enqueued orders as executed orders: Charge immediately and also don't need to consider lifetimes anymore.

Thinking about future proofing, I think some paging mechanism will likely be a good idea anyway. I'll create a ticket.

**Comment by eskimor** [comment]
One important thing to consider in any case: If the price cannot fluctuate withing a block, then safe economic bounds might not be possible. Issue: #6719

**Comment by rphmeier** [comment]
Are you concerned about a low price being exploited, by a block author providing a large amount of claims?

A maximum amount of "inflow" per block might be configurable if so, as long as it doesn't cause the transaction to disappear forever.

If prices can vary within a block, it gives the block author some flexibility over the ordering of parathread claims.

**Comment by burdges** [comment]
We can always charge the same price for every parathrad in a block, but a fuller block makes everyone pay more.  We'd have a problem if all this money went to the block producers, but we'll should just burn enough so they cannot benefit by wasting resources.

**Comment by eskimor** [comment]
> Are you concerned about a low price being exploited, by a block author providing a large amount of claims?

Indeed! 

> A maximum amount of "inflow" per block might be configurable if so, as long as it doesn't cause the transaction to disappear forever.

Yes you are right, this should work as well with the transaction validity checks! :100:

**Comment by rphmeier** [comment]
Yeah, also worth noting that "everyone in the block pays the same price" is not the same as "everyone in the block pays the price as-of the beginning of the block".

Take a number of bids B. Then `total_price = (0..B-1).fold(block_start_price, |a, _| a + next_price(a))`. Then you could charge everyone `total_price / B` - a type of logarithmic average price, assuming that `next_price` is an exponential function.

This is elegant but not easy to implement because it means computing the final price at the end & a lot of packing complexity for block authors. Probably better to keep it simple.
